File: vbox-app/src/main/java/com/kedzie/vbox/machine/TakeSnapshotFragment.java
package com.kedzie.vbox.machine;

import android.app.Dialog;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.TextView.OnEditorActionListener;


import com.kedzie.vbox.R;
import com.kedzie.vbox.api.IConsole;
import com.kedzie.vbox.api.IMachine;
import com.kedzie.vbox.api.IProgress;
import com.kedzie.vbox.api.ISnapshot;
import com.kedzie.vbox.app.BundleBuilder;
import com.kedzie.vbox.soap.VBoxSvc;
import com.kedzie.vbox.task.ActionBarTask;
import com.kedzie.vbox.task.MachineTask;
import roboguice.fragment.RoboDialogFragment;

/**
 * Create a new snapshot
 * 
 * @apiviz.stereotype fragment
 */
public class TakeSnapshotFragment extends RoboDialogFragment {
    
	private VBoxSvc _vmgr;
	private IMachine _machine;
	private ISnapshot _snapshot;
	
	private TextView _nameText;
	private TextView _descriptionText;
	
	public static TakeSnapshotFragment getInstance(VBoxSvc vmgr, IMachine machine,ISnapshot snapshot) {
	    BundleBuilder b = new BundleBuilder().putParcelable(VBoxSvc.BUNDLE, vmgr).putProxy(IMachine.BUNDLE, machine);
	    if(snapshot!=null)
	        b.putProxy(ISnapshot.BUNDLE, snapshot);
		TakeSnapshotFragment f = new TakeSnapshotFragment();
		f.setArguments(b.create());
		return f;
	}
	
	@Override
	public Dialog onCreateDialog(Bundle savedInstanceState) {
		Dialog dialog = super.onCreateDialog(savedInstanceState);
		dialog.setTitle(_snapshot==null ? getResources().getString(R.string.new_snapshot_dialog_title) : getResources().getString(R.string.edit_snapshot_dialog_title));
		return dialog;
	}

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		_vmgr = BundleBuilder.getProxy(getArguments(), VBoxSvc.BUNDLE, VBoxSvc.class);
		_machine = BundleBuilder.getProxy(getArguments(), IMachine.BUNDLE, IMachine.class);
		if(getArguments().containsKey(ISnapshot.BUNDLE)) 
		    _snapshot = BundleBuilder.getProxy(getArguments(), "snapshot", ISnapshot.class);
	}

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		View view = inflater.inflate(R.layout.snapshot_dialog, null);
		_nameText = (TextView)view.findViewById(R.id.snapshot_name);
		_descriptionText = (TextView)view.findViewById(R.id.snapshot_description);
		_descriptionText.setOnEditorActionListener(new OnEditorActionListener() {
			@Override
			public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
				if ((event != null && (event.getKeyCode() == KeyEvent.KEYCODE_ENTER)) || (actionId == EditorInfo.IME_ACTION_DONE)) {
					takeSnapshot();
	            }  
				return false;
			}
		});
		if(_snapshot != null) {
		    _nameText.setText(_snapshot.getName());
		    _descriptionText.setText(_snapshot.getDescription());
		}
		((ImageButton)view.findViewById(R.id.button_save)).setOnClickListener( new View.OnClickListener() {
			@Override
			public void onClick(View arg0) {
				dismiss();
				if(_snapshot!=null) {
    				new ActionBarTask<ISnapshot, Void>((AppCompatActivity)getActivity(), _vmgr) {
                        protected Void work(ISnapshot...s) throws Exception {     
                            s[0].setName(_nameText.getText().toString());
                            s[0].setDescription(_descriptionText.getText().toString());
                            return null;
                        }
                    }.execute(_snapshot);    
				} else {
    				takeSnapshot();
				}
			}
		});
		((ImageButton)view.findViewById(R.id.button_cancel)).setOnClickListener(new View.OnClickListener() { 
			public void onClick(View v) { 
				dismiss(); 
			} 
		});
		return view;
	}
	
	private void takeSnapshot() {
		new MachineTask<Void, Void>((AppCompatActivity)getActivity(), _vmgr, R.string.progress_taking_snapshot, false, _machine) {
			protected IProgress workWithProgress(IMachine m, IConsole console, Void...i) throws Exception { 	
				return console.takeSnapshot( _nameText.getText().toString(),  _descriptionText.getText().toString()); 
			}
		}.execute();
	}
}

File: vbox-app/src/main/java/com/kedzie/vbox/machine/TakeSnapshotFragment.kt
package com.kedzie.vbox.machine

import android.app.Dialog
import android.content.Intent
import android.os.Bundle
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import androidx.fragment.app.DialogFragment
import androidx.lifecycle.viewModelScope
import androidx.navigation.fragment.navArgs
import com.kedzie.vbox.R
import com.kedzie.vbox.VMAction
import com.kedzie.vbox.api.IMachine
import com.kedzie.vbox.api.IProgress
import com.kedzie.vbox.api.jaxb.LockType
import com.kedzie.vbox.api.jaxb.SessionState
import com.kedzie.vbox.soap.VBoxSvc
import com.kedzie.vbox.task.ProgressService
import kotlinx.android.synthetic.main.snapshot_dialog.*
import kotlinx.coroutines.launch
import org.koin.androidx.viewmodel.ext.android.sharedViewModel
import org.koin.core.parameter.parametersOf

/**
 * Create a new snapshot
 *
 * @apiviz.stereotype fragment
 */
class TakeSnapshotFragment : DialogFragment() {

    private val model: MachineListViewModel by sharedViewModel { activity!!.intent.let {
        parametersOf(it.getParcelableExtra(VBoxSvc.BUNDLE), it.getParcelableExtra(IMachine.BUNDLE)) } }

    private val args: TakeSnapshotFragmentArgs by navArgs()

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        return super.onCreateDialog(savedInstanceState).apply {
            setTitle(if (args.snapshot == null) resources.getString(R.string.new_snapshot_dialog_title) else resources.getString(R.string.edit_snapshot_dialog_title))
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.snapshot_dialog, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        snapshot_description.setOnEditorActionListener { v, actionId, event ->
            if (event != null && event.keyCode == KeyEvent.KEYCODE_ENTER || actionId == EditorInfo.IME_ACTION_DONE) {
                takeSnapshot()
            }
            false
        }
        args.snapshot?.let {
            model.viewModelScope.launch {
                snapshot_name.setText(it.getName())
                snapshot_description.setText(it.getDescription())
            }
        }

        button_save.setOnClickListener {
            dismiss()
            if (args.snapshot != null) {
                model.viewModelScope.launch {
                    args.snapshot?.setName(snapshot_name.text.toString())
                    args.snapshot?.setDescription(snapshot_description.text.toString())
                }
            } else {
                takeSnapshot()
            }
        }
        button_cancel.setOnClickListener { dismiss() }
    }

    private fun handleProgress(p: IProgress, action: VMAction) =
            activity?.startService(Intent(activity, ProgressService::class.java)
                    .putExtra(IProgress.BUNDLE, p)
                    .putExtra(ProgressService.INTENT_ICON, action.drawable()))

    private fun takeSnapshot() {
        model.vmgr.value?.vbox?.let { vbox ->
            model.viewModelScope.launch {
                val session = vbox.getSessionObject()
                if (session.getState() == SessionState.UNLOCKED)
                    model.machine.value!!.lockMachine(session, LockType.SHARED)
                try {
                    handleProgress(session.getConsole().takeSnapshot(snapshot_name.text.toString(), snapshot_description.text.toString()), VMAction.TAKE_SNAPSHOT)
                } finally {
                    if (session.getState() == SessionState.LOCKED)
                        session.unlockMachine()
                }
            }
        }
    }
}